File: csharp/ZeroBuffer/DuplexChannel/ImmutableDuplexServer.cs
using System;
using System.Threading;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace ZeroBuffer.DuplexChannel
{
    /// <summary>
    /// Server implementation that processes immutable requests and returns new response data
    /// </summary>
    internal class ImmutableDuplexServer : IImmutableDuplexServer
    {
        private readonly string _channelName;
        private readonly BufferConfig _config;
        private readonly TimeSpan? _timeout;
        private readonly ILogger<ImmutableDuplexServer> _logger;
        private Reader? _requestReader;
        private Writer? _responseWriter;
        private Thread? _processingThread;
        private CancellationTokenSource? _cancellationTokenSource;
        private volatile bool _isRunning;
        private bool _disposed;
        
        public event EventHandler<ErrorEventArgs>? OnError;
        
        public ImmutableDuplexServer(string channelName, BufferConfig config,TimeSpan? timeout=null,  ILogger<ImmutableDuplexServer>? logger = null)
        {
            _channelName = channelName;
            _config = config;
            _timeout = timeout;
            _logger = logger ?? NullLogger<ImmutableDuplexServer>.Instance;
        }
        
        public bool IsRunning => _isRunning;
        public static bool WaitExists(string channelName, TimeSpan timeout)
        {
            ArgumentException.ThrowIfNullOrEmpty(channelName);

            // For duplex channel, we need to check the request buffer
            // which is created by the server (with Reader)
            var requestBufferName = $"{channelName}_request";

            // Use the Reader's WaitExists to check if the request buffer exists
            return Reader.WaitExists(requestBufferName, timeout);
        }
        public void Start(RequestHandler onFrame, Action<ReadOnlySpan<byte>>? onInit = null, ProcessingMode mode = ProcessingMode.SingleThread)
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(ImmutableDuplexServer));
                
            if (_isRunning)
                throw new InvalidOperationException("Server is already running");
                
            if (onFrame == null)
                throw new ArgumentNullException(nameof(onFrame));
                
            // Channel naming convention:
            var requestBufferName = $"{_channelName}_request";
            var responseBufferName = $"{_channelName}_response";
            
            try
            {
                // Create request buffer as reader (we own this)
                _requestReader = new Reader(requestBufferName, _config);
                
                _cancellationTokenSource = new CancellationTokenSource();
                _isRunning = true;
                
                // Start processing based on mode
                if (mode == ProcessingMode.SingleThread)
                {
                    // Start processing thread - it will connect to response buffer when available
                    _processingThread = new Thread(() => ProcessRequests(onFrame, responseBufferName, onInit, _cancellationTokenSource.Token))
                    {
                        Name = $"DuplexServer_{_channelName}",
                        IsBackground = false
                    };
                    _processingThread.Start();
                }
                else
                {
                    // ThreadPool mode not yet implemented
                    throw new NotSupportedException($"Processing mode {mode} is not yet implemented");
                }
            }
            catch
            {
                Stop();
                throw;
            }
        }
        
        public void Stop()
        {
            if (!_isRunning)
                return;
                
            _isRunning = false;
            _cancellationTokenSource?.Cancel();
            
            // Wait for processing thread to complete
            _processingThread?.Join(TimeSpan.FromSeconds(5));
            
            // Clean up resources
            _cancellationTokenSource?.Dispose();
            _responseWriter?.Dispose();
            _requestReader?.Dispose();
            
            _cancellationTokenSource = null;
            _processingThread = null;
            _responseWriter = null;
            _requestReader = null;
        }
        
        public void Dispose()
        {
            if (_disposed)
                return;
                
            _disposed = true;
            Stop();
        }
        
        private Writer ConnectToResponseBuffer(string bufferName)
        {
            // Use the configured timeout, not a hardcoded value
            // Default to 5 seconds if no timeout specified
            var timeout = _timeout ?? TimeSpan.FromSeconds(5);
            int maxRetries = (int)(timeout.TotalMilliseconds / 100); // Convert to 100ms intervals
            
            for (int i = 0; i < maxRetries; i++)
            {
                try
                {
                    return new Writer(bufferName);
                }
                catch (BufferNotFoundException)
                {
                    Thread.Sleep(100);
                }
            }
            
            throw new TimeoutException($"Timeout waiting for response buffer {bufferName} after {timeout.TotalSeconds} seconds");
        }
        
        private void ProcessRequests(RequestHandler handler, string responseBufferName,
            Action<ReadOnlySpan<byte>> metadata, CancellationToken cancellationToken)
        {
            // Connect to response buffer when it becomes available
            if (_responseWriter == null!)
            {
                try
                {
                    _responseWriter = ConnectToResponseBuffer(responseBufferName);
                }
                catch (Exception ex)
                {
                    OnError?.Invoke(this, new ErrorEventArgs(ex));
                    _logger.LogError(ex, "Failed to connect to response buffer {BufferName}", responseBufferName);
                    _isRunning = false;
                    return;
                }
            }
            
            while (!cancellationToken.IsCancellationRequested && _isRunning)
            {
                try
                {
                    // Read request with timeout - using ensures Dispose is called
                    using var request = _requestReader.ReadFrame(_timeout);
                    if (!request.IsValid)
                        continue;

                    if (metadata != null)
                    {
                        metadata(_requestReader.GetMetadata());
                        metadata = null; // Only call once
                    }
                    // Let handler process request and write response directly
                    try
                    {
                        handler(request, _responseWriter);
                    }
                    catch (Exception ex)
                    {
                        // Log error but continue processing
                        OnError?.Invoke(this, new ErrorEventArgs(ex));
                        _logger.LogError(ex, "Error in request handler for channel {ChannelName}", _channelName);
                    }
                }
                catch (ReaderDeadException ex)
                {
                    OnError?.Invoke(this, new ErrorEventArgs(ex));
                    _logger.LogInformation("Reader disconnected", _channelName);
                    // Client disconnected
                    break;
                }
                catch (WriterDeadException ex)
                {
                    OnError?.Invoke(this, new ErrorEventArgs(ex));
                    _logger.LogInformation("Writer disconnected", _channelName);
                    break;
                }
                catch (Exception ex) when (!cancellationToken.IsCancellationRequested)
                {
                    // Log unexpected errors
                    OnError?.Invoke(this, new ErrorEventArgs(ex));
                    _logger.LogError(ex, "Server processing error on channel {ChannelName}", _channelName);
                }
            }
            _isRunning = false;
        }
    }
}

File: cpp/src/immutable_duplex_server.cpp
#include "zerobuffer/immutable_duplex_server.h"
#include "zerobuffer/logger.h"
#include <cstring>
#include <stdexcept>
#include <chrono>

namespace zerobuffer {

ImmutableDuplexServer::ImmutableDuplexServer(const std::string& channel_name, const BufferConfig& config)
    : channel_name_(channel_name)
    , config_(config)
{
}

ImmutableDuplexServer::~ImmutableDuplexServer() {
    stop();
}

void ImmutableDuplexServer::start(const ImmutableHandler& handler) {
    if (is_running_.exchange(true)) {
        throw std::runtime_error("Server is already running");
    }
    
    if (!handler) {
        throw std::invalid_argument("Handler cannot be null");
    }
    
    // Channel naming convention:
    const std::string request_buffer_name = channel_name_ + "_request";
    const std::string response_buffer_name = channel_name_ + "_response";
    
    ZEROBUFFER_LOG_INFO("ImmutableDuplexServer") << "Starting server on channel " << channel_name_;
    ZEROBUFFER_LOG_DEBUG("ImmutableDuplexServer") << "Request buffer: " << request_buffer_name;
    ZEROBUFFER_LOG_DEBUG("ImmutableDuplexServer") << "Response buffer: " << response_buffer_name;
    
    try {
        // Create request buffer as reader (we own this)
        request_reader_ = std::make_unique<Reader>(request_buffer_name, config_);
        
        should_stop_ = false;
        
        // Start processing thread - it will connect to response buffer when available
        processing_thread_ = std::thread([this, handler, response_buffer_name]() {
            process_requests(handler, response_buffer_name);
        });
    }
    catch (...) {
        stop();
        throw;
    }
}

void ImmutableDuplexServer::stop() {
    if (!is_running_.exchange(false)) {
        return;
    }
    
    should_stop_ = true;
    
    // Wait for processing thread to complete
    if (processing_thread_.joinable()) {
        processing_thread_.join();
    }
    
    // Clean up resources
    response_writer_.reset();
    request_reader_.reset();
}

bool ImmutableDuplexServer::is_running() const {
    return is_running_;
}

std::unique_ptr<Writer> ImmutableDuplexServer::connect_to_response_buffer(const std::string& buffer_name) {
    constexpr int max_retries = 50; // 5 second timeout
    constexpr auto retry_delay = std::chrono::milliseconds(100);
    
    for (int i = 0; i < max_retries; ++i) {
        try {
            return std::make_unique<Writer>(buffer_name);
        }
        catch (const std::exception&) {
            std::this_thread::sleep_for(retry_delay);
        }
    }
    
    throw std::runtime_error("Timeout waiting for response buffer " + buffer_name);
}

void ImmutableDuplexServer::process_requests(const ImmutableHandler& handler, const std::string& response_buffer_name) {
    ZEROBUFFER_LOG_INFO("ImmutableDuplexServer") << "Processing thread started for channel " << channel_name_;
    
    // Connect to response buffer when it becomes available
    if (!response_writer_) {
        try {
            ZEROBUFFER_LOG_DEBUG("ImmutableDuplexServer") << "Connecting to response buffer: " << response_buffer_name;
            response_writer_ = connect_to_response_buffer(response_buffer_name);
            ZEROBUFFER_LOG_INFO("ImmutableDuplexServer") << "Connected to response buffer";
        }
        catch (const std::exception& ex) {
            ZEROBUFFER_LOG_ERROR(channel_name_) << "Failed to connect to response buffer: " << ex.what();
            return;
        }
    }
    
    while (!should_stop_ && is_running_) {
        try {
            // Check if reader still exists (might be destroyed during stop)
            if (!request_reader_) {
                break;
            }
            
            // Read request with timeout
            ZEROBUFFER_LOG_TRACE("ImmutableDuplexServer") << "Waiting for request...";
            Frame request = request_reader_->read_frame(std::chrono::seconds(1));
            if (!request.is_valid()) {
                // Timeout or invalid frame - check if we should stop
                if (should_stop_) {
                    break;
                }
                continue;
            }
            ZEROBUFFER_LOG_DEBUG("ImmutableDuplexServer") << "Received request seq=" << request.sequence() << " size=" << request.size();
            
            // Let handler process request and write response directly
            try {
                handler(std::move(request), *response_writer_);
            }
            catch (const std::exception& ex) {
                // Log error but continue processing
                ZEROBUFFER_LOG_ERROR(channel_name_) << "Error in request handler: " << ex.what();
            }
        }
        catch (const std::exception& ex) {
            // Log unexpected errors unless we're shutting down
            if (!should_stop_) {
                ZEROBUFFER_LOG_ERROR(channel_name_) << "Server processing error: " << ex.what();
            } else {
                ZEROBUFFER_LOG_DEBUG("ImmutableDuplexServer") << "Server stopped: " << ex.what();
            }
        }
    }
}

} // namespace zerobuffer